### PR TITLE
[storage] state pruner: add mutex to protect pruning progress

### PR DESCRIPTION
**Pull request by msmouse**
### Description

Test no longer flaky on my dev machine.

### Test Plan

 ```
for i in `seq 1 30`; do cargo test -p aptosdb -- test_state_merkle_pruning; done
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3364)

### CODE REVIEW

File: aptos-move/framework/cached-packages/src/aptos_framework_sdk_builder.rs [hunk 1]
```diff
@@ -131,11 +131,6 @@ pub enum EntryFunctionCall {
         coin_type: TypeTag,
     },
```
Comment (zcchahaha) [inline aptos-move/framework/cached-packages/src/aptos_framework_sdk_builder.rs:132]: Why do we change this file?

File: storage/aptosdb/src/pruner/state_store/mod.rs [hunk 2]
```diff
@@ -92,24 +90,19 @@ where
...
     }
 
+    // used only by blanket `initialize()`, use the underlying implementation instead elsewhere.
     fn record_progress(&self, min_readable_version: Version) {
-        self.min_readable_version
-            .store(min_readable_version, Ordering::Relaxed);
-        PRUNER_LEAST_READABLE_VERSION
-            .with_label_values(&["state_store"])
-            .set(min_readable_version as i64);
+        self.record_progress_impl(min_readable_version, false);
```
Comment (zcchahaha) [inline storage/aptosdb/src/pruner/state_store/mod.rs:95]: nit: `/*is_fully_pruned=*/false`

File: storage/aptosdb/src/pruner/state_store/mod.rs [hunk 3]
```diff
@@ -92,24 +90,19 @@ where
...
-                .pruned_to_the_end_of_target_version
-                .load(Ordering::Relaxed)
+        let (min_readable_version, fully_pruned) = *self.progress.lock();
+        self.target_version() > min_readable_version || !fully_pruned
     }
 
     /// (For tests only.) Updates the minimal readable version kept by pruner.
     fn testonly_update_min_version(&self, version: Version) {
-        self.min_readable_version.store(version, Ordering::Relaxed)
+        self.record_progress_impl(version, true);
```
Comment (zcchahaha) [inline storage/aptosdb/src/pruner/state_store/mod.rs:105]: nit: `/*is_fully_pruned=*/true`

File: storage/aptosdb/src/pruner/state_store/mod.rs [hunk 4]
```diff
@@ -179,13 +169,18 @@ where
...
             // progress.
-            self.record_progress(new_min_readable_version);
-            self.pruned_to_the_end_of_target_version
-                .store(is_end_of_target_version, Ordering::Relaxed);
+            self.record_progress_impl(new_min_readable_version, is_end_of_target_version);
             Ok(new_min_readable_version)
         }
     }
 
+    fn record_progress_impl(&self, min_readable_version: Version, is_full_pruned: bool) {
```
Comment (zcchahaha) [inline storage/aptosdb/src/pruner/state_store/mod.rs:177]: nit: `is_full_pruned` -> `is_fully_pruned`

**Review by zcchahaha — APPROVED** [review]
Is the framework change intended?

File: aptos-move/framework/cached-packages/src/aptos_framework_sdk_builder.rs (hunk 1 above)
Comment (msmouse) [reply aptos-move/framework/cached-packages/src/aptos_framework_sdk_builder.rs:132]: See discussion in #move, a previous PR should have checked these in but somehow it got through without.

**Review by grao1991 — APPROVED** [review]

**Pull request merged**